crates/nu-parser/src/hir/path.rs

```diff
@@ -39,3 +39,105 @@ impl Path {
         (self.head, self.tail)
     }
 }
+
+// WIP: find a better place for these functions.
+
+use libc::getpwnam;
+use std::error::Error;
+use std::ffi::{CStr, CString, NulError};
+use std::fmt;
+use std::path::PathBuf;
+
+type HomedirResult = Result<Option<String>, UsernameError<NulError>>;
+
+/// Matches exising user names
+///
+/// Used to resolve `~username` syntax, may fail if the username contains a null with
+/// `NulError`. Otherwise will return a `String` with the user dir (not checked to be a
+/// valid path) or `None` if there is no matching username.
+pub(crate) fn username_homedir(name: &str) -> HomedirResult {
+    let cname = CString::new(name).map_err(|e| UsernameError {
+        username: name.into(),
+        cause: e,
+    })?;
+    fn get_dir_from_username(cname: CString) -> Option<String> {
+        let pwd = unsafe { getpwnam(cname.as_ptr()) };
```

Comment:
I used the non reentrant version of the function, I'd better switch to `getpwnam_r`

Reply:
I'm also investigating if we can use a minimal crate like https://github.com/sp3d/tilde-expand-rs to do the work, but the implementation has no dependency injection hooks like the `shellexpand` crate one, so the option to extract only the needed pieces is still open.